### ISSUE TITLE
Construct Multinomial conditional distributions, where total_count is a random variable

### ISSUE THREAD

**Issue by MariaLavrovskaya** [tensorflow/probability#1624]
I am trying to build a graphical model in Tensorflow Probability, where we first sample a number of positive (1) and negative (0) examples (count_i) from Categorical distribution and then construct Multinomial distribution (Y_i) depending on the value of (count_i). These events (Y_i) are mutually exclusive :

Y_1 ~ Multinomial([.9, 0.1, 0.05, 0.05, 0.1], total_count = [tf.reduce_sum(tf.cast(count==1, tf.float32))

Y_2 ~ Multinomial([0.99, 0.01, 0., 0., 0.], total_count = [tf.reduce_sum(tf.cast(count==0, tf.float32))

I have read these [tutorials](https://www.tensorflow.org/probability/examples/Multilevel_Modeling_Primer#53_varying_slopes), however I am stuck with two issues:

1. This code generates two arrays of length 500, whereas I only need 1 array of 500. What should I change so we only get 1 sample from Categorical distribution and then depending on the overall count of the value we are conditioning on, Multinomial is constructed ?
2. The sample from Categorical distribution gives only values of 0, whereas it should be a blend between 0 and 1. What am I doing wrong here?

My code is as follows. You can run these to replicate the behaviour:

```
    def simplied_model():
      return tfd.JointDistributionSequential([
          tfd.Uniform(low=0., high = 1., name = 'e'), #e
          
          lambda e: tfd.Sample(tfd.Categorical(probs = tf.stack([e, 1.-e], 0)), sample_shape =500), #count #should it be independent?
          lambda count: tfd.Multinomial(probs = tf.constant([[.9, 0.1, 0.05, 0.05, 0.1], [0.99, 0.01, 0., 0., 0.]]), total_count  = tf.cast(tf.stack([tf.reduce_sum(tf.cast(count==1, tf.float32)),tf.reduce_sum(tf.cast(count==0, tf.float32))], 0), dtype= tf.float32))
      ])

    tt = simplied_model()
    tt.resolve_graph()
    tt.sample(1)
```

**Comment by MariaLavrovskaya** [comment]
Published on stackoverflow

**Comment by csuter** [comment]
I think you just need to change the `axis` argument to `tf.stack` from `0` (left most) to `-1` (right most).

When you pass a `sample_shape` to `sample`, the output of `tfd.Uniform(..).sample(...)` will take that shape. In this case, with `sample_shape = 1`, the uniform has shape [1]. Then when you `tf.stack([e, 1-e], 0)`, you get shape `[2, 1]`. But what you want here is shape `[1, 2]`. Passing a shape `[2, 1]` probs array to Categorical will actually result in a batch of 2 independent, 1-category Categoricals -- we'll get 2 arrays out and all the samples will be 0! If we instead stack on axis -1, we'll get probs of shape `[1, 2]`, which has the right number of categories (innermost shape), and will yield a shape [500, 1] output, which we can then sum over axis 0 in the last distribution to get the right output.

Hope this makes some sense!

**Comment by MariaLavrovskaya** [comment]
Thank you so much. Totally makes sense! Should have spotted it myself

**Comment by MariaLavrovskaya** [comment]
Apologies for bothering again, but I am experiencing issues getting the dims right for the Dirichlet Multinomial model and to generalise it so I can run MCMC on it. I've managed to build the function for one sample, but I cannot seem to figure out how to generalise it to the batched version. The idea is that every time we generate a sample, we want to build two Multinomial distributions. I cannot seem to figure out what the dimensions supposed to look like for posterior. I have tried adding an axis but it does not seem to help.

I have also tried printing out everything and I cannot understand why the dimensions first are for one sample but they it complains. When we run sample, it is vectorised right? So both of distributions for each of the samples should be built at the same time?

My understanding is that for the sample of 2, we need the priors to be of the shape (2,2,5) and counts of the shape (2,2,), however I think I am missing something because it does not work. Any hints please?
```
#means = [0.0623815 , 0.4076164 , 0.49531835, 0.01912234, 0.0155614 ],
#bac = [0.9 , 0.1 , 0.05, 0.05, 0.1 ]
```
````
def _get_probs(priors):
  answer = tf.stack([tf.reshape(priors, [5]), tf.constant([.9, 0.1, 0.05, 0.05, 0.1], dtype = tf.float32)], 0)
  print(priors.shape, 'pass', answer.shape)
  return answer
def _get_counts(vec):
  answer = tf.cast(tf.stack([tf.reduce_sum(tf.cast(vec==1, tf.float32),0),tf.reduce_sum(tf.cast(vec==0, tf.float32),0)], 0), dtype= tf.float32)
  print('pass', answer.shape)
  return answer
def simplified_model(means, bac):
  return tfd.JointDistributionSequential([
      
      tfd.Uniform(low=0., high = 1.), #b
      lambda b: tfd.Sample(tfd.Categorical(probs = tf.stack([b, 1.-b], -1)), sample_shape =[500], name = 'count'), #count
                tfd.Dirichlet(concentration = means, name = 'priors'), #priors
      lambda priors,count:  tfd.Multinomial(probs = _get_probs(priors), total_count  = tf.cast(tf.stack([tf.reduce_sum(tf.cast(count==1, tf.float32)),tf.reduce_sum(tf.cast(count==0, tf.float32))], 0), dtype= tf.float32)) #we can replace this with categorical distribution
     
  ])`
```

**Comment by csuter** [comment]
Couple of things going on here:
  1. JointDistributions have to do a bit of introspection the first time they are used, so that they know what shapes and dtypes and such to expect. So the forward sampling process is run once with sample_shape [1] (or maybe scalar, []?) to do this "type inference" and the result stored. This is why you see your helpers called with trivial shape
  2. The reshape, as written won't work on the non-trivial sample_shape forward pass -- your array has shape [2, 5] and can't be reshaped into shape [5]. You have at least 2 options:
      a. rewrite the helper function logic to account for additional leading batch dimensions
      b. use JointDistributionSequentialAutoBatched

2b is easiest, but it was pretty slow when I ran it. This is probably due to Multinomial being...pretty terrible! JDSAutoBatched uses tf.vectorized_map to do some auto-batching magic, which involves some compilation/program transformation that may be ill-suited to Multinomial's implementation. That said, it may be hard to do better vectorizing by hand. I'd suggest trying both (feel free to pop back in here w/ questions!), and see which is best for your purposes.

HTH!

**Comment by MariaLavrovskaya** [comment]
Thank you very much for being so supportive. I decided to go for not the easiest version first. I've managed to reshape the probs and the counts to pass it to Multinomial so it generate samples bigger than 1. However, I have a number of concerns: 

Right now, for the scenario with sample_shape = 2, I have probs of shape = (4,1,5) and associated counts (4,), which creates four different Multinomials as requested. However, I think that it's actually supposed to be (2,2,5) for probs and (2,2) for counts, but I do not see how to transform it in such a way for now. Do you think leaving shapes as they are now would affect MCMC in the future?
```
def _get_probs_v2(means, priors):
  return tf.concat([priors, non_claim_means], 0)`
```
```

def _get_counts(count, priors):
  multiple = tf.cast(tf.shape(priors)[0], dtype = tf.int32)
  count = tf.reshape(count, [multiple, 500]) #500 will be RV in itself 
  return tf.reshape(tf.stack([tf.reduce_sum(tf.cast(count==1, tf.float32), axis = 1),tf.reduce_sum(tf.cast(count==0, tf.float32), axis = 1)], -1), [-1])
```

```
def simplified_model(means_2, means):
  return tfd.JointDistributionSequential([
      
      # tfd.Dirichlet(concentration = means), #priors
      tfd.Uniform(low=0., high = 1.), #b
      lambda b: tfd.Sample(tfd.Categorical(probs = tf.stack([b, 1.-b], -1)), sample_shape =[500], name = 'count'), #count
                tfd.Dirichlet(concentration = means, name = 'priors'), #priors
                tfd.Dirichlet(concentration = means_2, name = 'means_b'),
      lambda means_b, priors, count:  tfd.Multinomial(probs = _get_probs_v2(means_b, priors), total_count  =_get_counts(count, priors)) 

```

**Comment by MariaLavrovskaya** [comment]
Thank you for being so supportive – it is much appreciated, especially because tensorflow is a bit hard to wrap my mind around it!

I have decided to go with the reshape first, because I thought it would be a good exercise to understand shapes better. As far as I understand for `sample_size > 1` (let's take 2), the final shape of the Multinomial is supposed to be `(2,2,5)` -> 4 distributions in total with two for each sample. Probs should be of size `(2,2,5)` and total counts should be of size `(2,2)`. It's definitely not the most elegant way to go about it, but it's definitely a step forward.

Will now look at the 2b as suggested and try running MCMC for both.

```
def _get_probs_v2(non_claim_means, priors):
  return tf.concat([priors, non_claim_means], 1) #changed from 0
def _get_counts(count, priors):
  multiple = tf.cast(tf.shape(priors)[0], dtype = tf.int32)
  count = tf.reshape(count, [multiple, 500]) #500 will be RV in itself 
  return tf.stack([tf.reduce_sum(tf.cast(count==1, tf.float32), axis = 1),tf.reduce_sum(tf.cast(count==0, tf.float32), axis = 1)], -1)

def simplified_model(means_2, means):
  return tfd.JointDistributionSequential([
      tfd.Uniform(low=0., high = 1.), #b
      lambda b: tfd.Sample(tfd.Categorical(probs = tf.stack([b, 1.-b], -1)), sample_shape =[500], name = 'count'), #count
                tfd.Dirichlet(concentration = means, name = 'priors'), #priors
                tfd.Dirichlet(concentration = means_2, name = 'means_b')
      lambda means_b, priors, count:  tfd.Multinomial(probs = _get_probs_v2(means_b, priors), total_count  =_get_counts(count, priors))
```

**Comment by brianwa84** [comment]
FWIW, if numpy is more familiar, you could use the same TFP API but in jax
instead.
from tensorflow_probability.substrates import jax as tfp

**Comment by MariaLavrovskaya** [comment]
Thank you both for your help! 

I am trying to run the MCMC sampler based on the run function [shown here](https://www.tensorflow.org/probability/examples/Modeling_with_JointDistribution#a_common_run_chain_function). I have adapted my code and found appropriate unconditional bijectors to map initial state. 

However, the algorithm fails because we are getting NaN values now and then. I am very confused and this time I have no clue if there is a workaround not to see NaN values. Am I doing something wrong or is there a trick to overcome this? 

The logic is as follows. Beta distribution controls for the proportion of 0 values in the generated sample. Dirichlet for 'first seen' controls the count for each of possible values occurring in 'seen' sample and Dirichlet for 'already seen' controls the sample of each of 5 possible values occurring in 'already seen' sample. They both have hyperparams coming from the data. 

```
def _get_probs_v2(already_seen, priors):
  print('prior_shape', priors.shape)
  return tf.concat([priors, already_seen], 1) #changed from 0

def _get_counts(count, priors):
  multiple = tf.cast(tf.shape(priors)[0], dtype = tf.int32)
  count = tf.reshape(count, [multiple, 50000]) 
  return tf.stack([tf.reduce_sum(tf.cast(count==1, tf.float32), axis = 1),tf.reduce_sum(tf.cast(count==0, tf.float32), axis = 1)], -1)

def simplified_model(first_seen, already_seen):
  return tfd.JointDistributionSequential([
      
      tfd.Beta(concentration1 = 559.49,concentration0= 291.28), #b
      lambda b: tfd.Sample(tfd.Categorical(probs = tf.stack([b, 1.-b], -1)), sample_shape =[50000], name = 'count'), #count
                tfd.Dirichlet(concentration = first_seen, name = 'priors'), #priors
                tfd.Dirichlet(concentration = already_seen, name = 'already_seen'), 
      lambda already_seen, priors, count:  tfd.Independent(tfd.Multinomial(probs = _get_probs_v2(already_seen, priors), total_count  =_get_counts(count, priors)), reinterpreted_batch_ndims=2)  #working_function
  ])
@tf.function
def simplified_log_prob(init_beta, init_first, init_already):
  return simplified_model(first_seen, already_seen).log_prob([init_beta, init_first, init_already, input_excess])
```
```
@tf.function
def sample_pooled(num_chains, num_results, num_burnin_steps, num_observations):
  """Samples from the pooled model."""
  hmc = tfp.mcmc.HamiltonianMonteCarlo(
      target_log_prob_fn=pooled_log_prob,
      num_leapfrog_steps=10,
      step_size=0.005)

  initial_state = [
      tf.zeros([num_chains], name='init_alpha'),
      tf.zeros([num_chains], name='init_beta'),
      tf.ones([num_chains], name='init_sigma')
  ]

  # Constrain `sigma` to the positive real axis. Other variables are
  # unconstrained.
  unconstraining_bijectors = [
      tfb.Identity(),  # alpha
      tfb.Identity(),  # beta
      tfb.Exp()        # sigma
  ]
  kernel = tfp.mcmc.TransformedTransitionKernel(
      inner_kernel=hmc, bijector=unconstraining_bijectors)

  samples, kernel_results = tfp.mcmc.sample_chain(
      num_results=num_results,
      num_burnin_steps=num_burnin_steps,
      current_state=initial_state,
      kernel=kernel)

  acceptance_probs = tf.reduce_mean(
      tf.cast(kernel_results.inner_results.is_accepted, tf.float32), axis=0)

  return samples, acceptance_probs
```

```
first_seen = tf.constant([[0.06225592, 0.40752962, 0.4955014 , 0.01913294, 0.01558013]], dtype = tf.float32)
already_seen = tf.constant([[0.9 , 0.1 , 0.05, 0.05, 0.1 ]], dtype = tf.float32)
```

Thank you again for being very supportive.

**Comment by csuter** [comment]
Can you add `allow_nan_stats=False` to all distribution constructors, and `validate_args=True` to all distribution and bijector constructors, and run again? You may get more informative errors and/or find source of NaNs sooner (I haven't had a close look at the latest code yet)

**Comment by csuter** [comment]
(no need to add these to args of tfd.Sample or tfd.Independent -- just the "concrete" distriubtions these are wrapping)

**Comment by MariaLavrovskaya** [comment]
Based on the first error that I am getting, the issue seems to be in the first helper function. Will try getting to the bottom of it.

**Comment by csuter** [comment]
Hmm. I think you can rewrite the 2-category Categorical + sum to be a tfd.Binomial(total_count=50000).

I'm confused by the log_prob call in simplified_log_prob because it is only passing 4 values but the JointDistribution has 5 random variables (beta, sample(categorical), dir1, dir2, multinomial).

If you can provide a runnable example in a Colab, I can try and debug more proactively.

**Comment by MariaLavrovskaya** [comment]
Yes, sure. https://colab.research.google.com/drive/1vSTW49QrkdIC-ciELe8nZ98N8tZefJSQ?usp=sharing

It seems like the errors are coming from Dirichlet distributions if I use` log_prob()` without rewriting it.

**Comment by MariaLavrovskaya** [comment]
I first rewrote the` simplified_log_prob()` as mentioned above because I thought for MCMC we need to define a closure over our joint to output prob for our data. However, I took a step back in this example and ran predefined `log_prob()` to trace the error step by step.

**Comment by MariaLavrovskaya** [comment]
Update: the `inf` issue is now sorted. Was incorrectly parametrising Dirichlet distributions with proportions, whereas they take mean counts. Trying to figure out how to rewrite the `log_prob` manually as the one out of box fails as it is expecting shape (n,n), whereas we need (n,2), n being the number of generated samples. Will report my progress along the way

**Comment by MariaLavrovskaya** [comment]
Almost giving up :( I slightly simplified the function to use Binomial instead of a Categorical (it's a version 2) and right now `log_prob()` seems to be working for the `sample_size` up to 2, but it starts failing for samples larger than 2. I am not sure if I can change the params somewhere so I would not need to write a `log_prob()` from scratch as I was trying not to use a handwritten approach from the very start. Would you mind having a look if you have time please? I'd greatly appreciate any suggestions! https://colab.research.google.com/drive/1vSTW49QrkdIC-ciELe8nZ98N8tZefJSQ?usp=sharing

**Comment by csuter** [comment]
I believe this will do what you want:

```python
def simplified_model_v2(first_seen, already_seen):
  return tfd.JointDistributionSequential([
    tfd.Beta(concentration1 = 559.49,
             concentration0= 291.28), #b
    lambda b: tfd.Binomial(probs=b, total_count=50000, name='count'),
    tfd.Independent(
        tfd.Dirichlet(concentration=tf.stack([first_seen, already_seen], axis=0)),
        reinterpreted_batch_ndims=1),
    lambda t, count: tfd.Independent(
        tfd.Multinomial(
            probs=t,
            total_count=tf.stack([count, 50000 - count], axis=-1)),
        reinterpreted_batch_ndims=1)
  ])
```

There are a couple things I needed to fix
  1. We can't just sub the Binomial with [p, 1-p] for what you were doing with the Categorical + sum. Since we want the total positive and negative trials to sum to 50000, we need to draw c from a Binomial(p, 50000) and use (50000 - c) for the negative count.
  2. I had to fiddle with shapes of things a lot...there isn't a great general solution to this TBH. The strategies include some combination of (a) "think really hard about shape semantics in the presence of non-trivial sample_shape", (b) "factor out the lambdas and add a bunch of print statements showing the shapes of intermediate values", and (c) "use JDSAutoBatched, which is sometimes painfully slow but was designed to save us from (a) and (b)".

Sorry about the challenges! I hope the next step is easier. I admire your perseverance trying to make it work w/out the AutoBatched magic :) It's also my personal preference, even though it's sometimes painful.

**Comment by MariaLavrovskaya** [comment]
Morning! Apologies I was taking a break over the weekend. This works smoothly – thank you so much. 

I've noticed error in (1) later on Friday and it was stupid of me to assume that it would work any differently. 

I cannot agree with you more. Although it's definitely sometimes painful and frustrating, but this is the best way to learn (as long as I can pop in here now and then to ask questions). 😸

**Comment by MariaLavrovskaya** [comment]
I think I managed to make MCMC work. I took values of simplified_model_v2(first_seen, already_seen).sample() to be initial_state and then imposed bijectors to map to unconstrained space (hopefully I did it correctly). I also wrote a partial application for the target log prob function

```
@tf.function
def simplified_log_prob(init_beta, init_binom, init_dir):
  return simplified_model_v2(first_seen, already_seen).log_prob([init_beta, init_binom, init_dir, inp_exc])
```
def sample_simplified(num_results, num_burnin_steps):
```
  hmc = tfp.mcmc.HamiltonianMonteCarlo(
      target_log_prob_fn=simplified_log_prob,
      num_leapfrog_steps=10,
      step_size=0.005)

  initial_state = [
      tf.constant(init_b, dtype=tf.float32, name='init_beta'),
      tf.constant(init_binom, name = 'init_binomial'),
      tf.constant(init_dirichlets, name = 'init_dir')
      # means* tf.ones([], dtype=tf.float32, name="init_means")
  ]

  unconstraining_bijectors = [
      tfb.Sigmoid(), #bijector for Beta 
      tfb.Identity(), #bijector for Binomial
      tfb.SoftmaxCentered() # Bijector for Dirichlets
  ]
  kernel = tfp.mcmc.TransformedTransitionKernel(
      inner_kernel=hmc, bijector=unconstraining_bijectors)

  samples, kernel_results = tfp.mcmc.sample_chain(
      num_results=num_results,
      num_burnin_steps=num_burnin_steps,
      current_state=initial_state,
      kernel=kernel)
  acceptance_probs = tf.reduce_mean(
      tf.cast(kernel_results.inner_results.is_accepted, tf.float32), axis=0)

  return samples, acceptance_probs
```

**Comment by MariaLavrovskaya** [comment]
Thank you so much for being so supportive and guiding me! 

I now need to introduce more complexity into the model by imposing another distribution for the `total_count `for Binomial, but hopefully I will manage it work smoothly myself 😅

**Comment by MariaLavrovskaya** [comment]
I decided to take a step forward and try to generalise it to fit a number of values at a time. Right now, we are fitting only one value at a time, however what I want is to fit 7 values at a time. 

So based on the explanations [here](https://www.tensorflow.org/probability/examples/JointDistributionAutoBatched_A_Gentle_Tutorial) and your valuable comments, I am trying to think about shapes really hard. My understanding is as follows: 

I started off by thinking that there are two possible scenarios:
Number one: 
1. Beta ~ two values (2,)
2. Binomial (2,)
3. Dirichlets (2,2,5)
4. Multinomials final shape is (2,7,2,5). 
For sample of two, it should output two values for `log_prob`. This looks like what I need, however I am not sure how I am supposed to transform the probs in the way so that it works as desired. My way of thinking was to add a dimension `t[:, tf.newaxis, ...]` but this fails for the forward pass `[2,5]`. 

Number two, which I now understand is wrong was to do the following: 
1. Beta ~ two values (2,)
2. Binomial (2,)
3. Dirichlets (2,2,5)
4. Multinomials final shape is (7,2,2,5).
Again, for sample of two, it should output two values for log_prob, but here I only managed to get one value for `log_prob` so this is probably not the right way to go forward. 

Do you think my understanding is correct and (1) is the right shapes I am supposed to be getting? Can you please suggest if my logic is right how can I formulate probs so that it works for sample_size greater than 1?

I put everything in the notebook to make it a bit more readable: https://colab.research.google.com/drive/1vSTW49QrkdIC-ciELe8nZ98N8tZefJSQ?usp=sharing

**Comment by brianwa84** [comment]
Instead of summing a bunch of two way categoricals, consider using Binomial.

**Comment by MariaLavrovskaya** [comment]
No worries. I hope you will meet your deadline!  I believe it works as intended 😸 I have another question that is slightly different to the thread's topic. Should I create a separate ticket for that?

I have seen in the[ tutorials](https://www.tensorflow.org/probability/examples/Bayesian_Switchpoint_Analysis) that there is a two Poisson model which two different rate parameters, is there a way to do this with two distributions from exponential family but which are parametrised differently? I am not sure if I am even allowed to do that. I believe I have a model which has LogNormal body and after certain 'switchpoint' it has exponential tail. So far I thought of using 'Mixture', however I am worried that it's not going to give me this 'switchpoint' behaviour. Any suggestions are very much welcomed!

**Comment by brianwa84** [comment]
No worries, I sent that response on Sep 26 by email, for some reason it only just showed up.

**Comment by MariaLavrovskaya** [comment]
Hi. Would you mind giving me advice on the bijector here? I want to create a mixture of two normals, where all the parameters are dictated by Uniform distributions. I am using three data arrays at a time, because I want to find parameters that would fit all of them. I am using this joint distribution and` log_prob` and `sample` work. I also adapted the bijector from the example to map the interval to unconstrained space, however Monte Carlo algorithm is not doing any steps. What might be wrong here?

```
def affine(rv):

  return tf.transpose(rv* tf.ones([3])[..., tf.newaxis])



def model_normal_normal():
  return tfd.JointDistributionSequential([
      tfd.Uniform(low = 0.08, high = 0.11), # weights
      tfd.Uniform(low = 4.8, high =  5.1), #loc_body
      tfd.Uniform(low = 6.4, high = 7.3), #loc_tail
      tfd.Uniform(low = 0.08, high = 0.1), #scale_body
      tfd.Uniform(low = 0.1, high = 0.2), #scale_tail
      lambda scale_tail, scale_body, loc_tail, loc_body, weights: tfd.Independent(tfd.Sample(tfd.Mixture(
          cat = tfd.Categorical(
              probs = tf.stack([affine(1-weights), affine(weights)], axis=-1), validate_args=True,allow_nan_stats=False),
              components = [
                  tfd.Normal(loc = affine(loc_body), scale = affine(scale_body), validate_args=True,allow_nan_stats=False), 
                  tfd.Normal(loc = affine(loc_tail), scale = affine(scale_tail), validate_args=True,allow_nan_stats=False)]), sample_shape = 23507),reinterpreted_batch_ndims=2)
      
  ])

def tfp_interval(a,b):
  return tfb.Inline(
    inverse_fn=(
        lambda x: tf.math.log(x - a) - tf.math.log(b - x)),
    forward_fn=(
        lambda y: (b - a) * tf.sigmoid(y) + a),
    forward_log_det_jacobian_fn=(
        lambda x: tf.math.log(b - a) - 2 * tf.nn.softplus(-x) - x),
    forward_min_event_ndims=0,
    name="interval")

def sample_normal_normal(num_results, num_burnin_steps):
  hmc = tfp.mcmc.HamiltonianMonteCarlo(
      target_log_prob_fn=log_prob_normal,
      num_leapfrog_steps=10,
      step_size=0.005)

  initial_state = [
      tf.constant(init_w, dtype=tf.float32, name='init_weights'),
      tf.constant(init_l_b, dtype=tf.float32, name = 'loc_body'),
      tf.constant(init_l_t, name = 'loc_tail'),
      tf.constant(init_s_b, name = 'scale_body'),
      tf.constant(init_s_t, name = 'scale_tail')
      # means* tf.ones([], dtype=tf.float32, name="init_means")
  ]
#we specify bijectors to map the initial values to unconstrained space
  unconstraining_bijectors = [
      tfp_interval(tf.constant(0.08, tf.float32), tf.constant(0.11, tf.float32)), #bijector for all uniforms
      tfp_interval(tf.constant(4.8, tf.float32), tf.constant(5.1, tf.float32)), 
      tfp_interval(tf.constant(6.4, tf.float32), tf.constant(7.3, tf.float32)),
      tfp_interval(tf.constant(0.08, tf.float32), tf.constant(0.1, tf.float32)),
      tfp_interval(tf.constant(0.1, tf.float32), tf.constant(0.2, tf.float32)) 
  ]
  kernel = tfp.mcmc.TransformedTransitionKernel(
      inner_kernel=hmc, bijector=unconstraining_bijectors)
  samples, kernel_results = tfp.mcmc.sample_chain(
      num_results=num_results,
      num_burnin_steps=num_burnin_steps,
      current_state=initial_state,
      kernel=kernel)
  acceptance_probs = tf.reduce_mean(
      tf.cast(kernel_results.inner_results.is_accepted, tf.float32), axis=0)

  return samples, acceptance_probs
```

**Comment by MariaLavrovskaya** [comment]
Thank you! It works with the smaller step size and `Sigmoid` bijector. Have another question: I was playing around with `reinterpreted_batch_ndims` and I do not seem to understand the difference. 
For example, if I sample once, each uniform produces 1 value. Then the mixture constructs three gaussian mixtures with roughly 20000 values each. When I set `reinterpreted_batch_ndims=1`, I have `batch_shape = [1]` and `event_shape = [3,20000]`, whereas when I set `reinterpreted_batch_ndims=2` I have `batch_shape = []` and `event_shape = [1,3,20000]`. Both work when `sample` and `log_prob` is called. Values in `log_prob` would differ for different `reinterpreted_batch_ndims`. However, the sampler only seems to work when `reinterpreted_batch_ndims=2` and throws an error when `reinterpreted_batch_ndims=1`. 

Would you mind explaining what is the right approach here and why this behaviour is intended?

**Comment by MariaLavrovskaya** [comment]
Found the answer to my question in the walk-through on shapes – thanks for help!

**Comment by csuter** [comment]
I don't think concentration is supposed to be negative?

**Comment by MariaLavrovskaya** [comment]
I figured. Concentration can be negative, however all values in x are bounded from above by `loc-scale/concentration`, hence the error.

**Comment by MariaLavrovskaya** [comment]
Hello! Is there an alternative in TFP to the posterior predictive samples from a model given a trace as in https://docs.pymc.io/en/v3/api/inference.html ?

**Comment by csuter** [comment]
The `sample` function on JointDistriubtions accepts arguments that let you "pin" certain values during sampling. you can use this to get posterior predictive samples, given posterior samples. The way we generate docs makes it a bit hard to suss out, but it's expressed most clearly (perhaps still not all that clearly...) in the `sample_distributions` docstring [here](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/JointDistribution#sample_distributions). The same applies to the sample method, even though the docstrings don't say so (they're inherited with modifications from the base Distribution object...😬)

In short you can do things like this:

```python
jds = tfd.JointDistributionSequential([
  tfd.Normal(0., 1., name='x')
  lambda x: tfd.Normal(x, 1., name = 'y')
])

# (1) pin value of x, but not y:
jds.sample(10, value=[4, None])

# (2) pin value of y, but not x:
jds.sample(10, value=[None, -3])

# (3) same as (1), but use kwargs to pin, based on the `name` parameters of the constituent distributions:
jds.sample(10, x=4)

# (4) same as (2), but use kwargs to pin
jds.sample(10, y=-3)
```

constructing a valid `value` argument can be a bit of a hassle (have to plug in `None`s in just the right places), hence the latter form.

**Comment by MariaLavrovskaya** [comment]
I see, many thanks!
